src/backend/distributed/planner/distributed_planner.c

```diff
@@ -2590,3 +2615,122 @@ WarnIfListHasForeignDistributedTable(List *rangeTableList)
 		}
 	}
 }
+
+
+/*
+ * AreTablesMergeSqlCommandSupported returns true if all the tables(target, source or any CTE
+ * present) in the MERGE command are local i.e. a combination of Citus local and Non-Citus
+ * tables (regular Postgres tables). Returns false for all other combinations.
+ */
+static bool
+AreTablesMergeSqlCommandSupported(List *rangeTableList)
+{
+	ListCell *tableCell = NULL;
+
+	foreach(tableCell, rangeTableList)
+	{
+		RangeTblEntry *rangeTableEntry = (RangeTblEntry *) lfirst(tableCell);
+		Oid relationId = rangeTableEntry->relid;
+
+		switch (rangeTableEntry->rtekind)
+		{
+			case RTE_RELATION:
+			{
+				/* Check the relation type */
+				break;
+			}
+
+			case RTE_SUBQUERY:
+			case RTE_CTE:
+			{
+				/* Skip the subquery and cte as the base table(s) will be checked */
+				continue;
+			}
+
+			case RTE_JOIN:
+			{
+				ereport(ERROR,
+						(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+						 errmsg("MERGE command is not supported with JOIN")));
```

Comment:
are these RTEs extracted only from the top-level merge, or also from subqueries? 

We should aim for full support of MERGE for local tables.

Reply:
I've added the support along with the unit tests, please take a look.